support struct array in pretty display

houqp opened the pull request:
# Which issue does this PR close?

support struct array in pretty print

# Rationale for this change
 
support display of nested fields in downstream projects that displays record batches in console.

# What changes are included in this PR?

code to format struct arrays in pretty print
print null value as null string to be consistent with postgresql

# Are there any user-facing changes?

better pretty print

alamb reviewed (commented):
Is there any way to add some test coverage to this code so can avoid introducing regressions (aka breaking your code in the future)?

houqp commented:
Yep, I will add some tests. I was thinking about this for the arrow2 migration as well ;)

houqp commented:
@alamb added struct test, also changed null display to be consistent with postgres and mysql.

alamb commented on arrow/src/util/pretty.rs:
```diff
@@ -507,4 +507,63 @@ mod tests {
+            r#"+-------------------------------------+----+"#,
+            r#"| c1                                  | c2 |"#,
+            r#"+-------------------------------------+----+"#,
+            r#"| {"c11": 1, "c12": {"c121": "e"}}    | a  |"#,
```
this is cool -- it is almost like JSON

alamb reviewed (commented):
@houqp 

THe struct tests look good 👍  thank you



> also changed null display to be consistent with postgres / mysql / sqlite.

I am not sure the string`"null"` is consistent with postgres. Here is what happens for me with `psql`:


```sql
alamb=# create table bar(x int);
CREATE TABLE
alamb=# insert into bar values (1);
INSERT 0 1
alamb=# insert into bar values (null);
INSERT 0 1
alamb=# select * from bar;
 x 
---
 1
  
(2 rows)

```


Sqlite also uses an empty string:

```shell
(arrow_dev) alamb@MacBook-Pro:~/Software/influxdb_iox2$ sqlite3 
SQLite version 3.32.3 2020-06-18 14:16:19
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
sqlite> create table bar(x int);
sqlite> insert into bar values (1);
sqlite> insert into bar values (null);
sqlite> select * from bar;
1

sqlite> 
```


mysql seems to use the string `"NULL"` rather than `"null"`:
```
mysql> use foo;
Database changed
mysql> create table bar(x int);
Query OK, 0 rows affected (0.00 sec)

mysql> insert into bar values (1);
Query OK, 1 row affected (0.01 sec)

mysql> insert into bar values (null);
Query OK, 1 row affected (0.01 sec)

mysql> select * from bar;
+------+
| x    |
+------+
|    1 |
| NULL |
+------+
2 rows in set (0.00 sec)
```

houqp commented:
> I am not sure the string"null" is consistent with postgres. Here is what happens for me with psql:

Good call, I was using a different postgres client to test the display. I have reverted that commit and updated the code to handle null only within struct display code.

alamb reviewed (approved):
Looks great @houqp  -- thanks!

jorgecarleitao commented on arrow/src/util/pretty.rs:
```diff
@@ -507,4 +507,63 @@ mod tests {
+            r#"| c1                                  | c2 |"#,
+            r#"+-------------------------------------+----+"#,
+            r#"| {"c11": 1, "c12": {"c121": "e"}}    | a  |"#,
+            r#"| {"c11": null, "c12": {"c121": "f"}} | b  |"#,
```
it could have been useful to have a struct with a validity tested, to make sure we do not miss that case.

houqp replied:
will send a follow up PR

jorgecarleitao commented on arrow/src/util/pretty.rs:
```diff
@@ -507,4 +507,63 @@ mod tests {
+            r#"| c1                                  | c2 |"#,
+            r#"+-------------------------------------+----+"#,
+            r#"| {"c11": 1, "c12": {"c121": "e"}}    | a  |"#,
+            r#"| {"c11": null, "c12": {"c121": "f"}} | b  |"#,
```
should we have a `null` in the fields' values but not in all other types?

houqp replied:
I don't have a strong opinion on this. I went with this way because I feel like it looks more clear than `"c11": ,`. There is also the other option of skipping the column with null value entirely in the display. Let me know if you feel strongly on one way or the other, happy to send a PR to change the behavior as well.
